scio-core/src/main/scala/com/spotify/scio/ScioResult.scala

```diff
@@ -20,52 +20,41 @@ package com.spotify.scio
 import java.nio.ByteBuffer
 
 import com.spotify.scio.metrics._
-import com.spotify.scio.options.ScioOptions
 import com.spotify.scio.util.ScioUtil
 import com.twitter.algebird.Semigroup
-import org.apache.beam.runners.dataflow.DataflowPipelineJob
-import org.apache.beam.runners.dataflow.options.DataflowPipelineOptions
 import org.apache.beam.sdk.Pipeline.PipelineExecutionException
 import org.apache.beam.sdk.PipelineResult.State
-import org.apache.beam.sdk.options.ApplicationNameOptions
-import org.apache.beam.sdk.PipelineResult
 import org.apache.beam.sdk.io.FileSystems
-import org.apache.beam.sdk.{metrics => bm}
 import org.apache.beam.sdk.util.MimeTypes
-import org.slf4j.LoggerFactory
+import org.apache.beam.sdk.{PipelineResult, metrics => bm}
 
 import scala.collection.JavaConverters._
-import scala.concurrent.{Await, Future}
 import scala.concurrent.duration.Duration
-import scala.util.control.NonFatal
-import scala.util.{Failure, Success, Try}
+import scala.concurrent.{Await, Future}
+import scala.reflect.ClassTag
+import scala.util.Try
+
+/** Represent a Beam runner specific result. */
+trait RunnerResult {
+  /** Get a generic [[ScioResult]]. */
+  def asScioResult: ScioResult
+}
 
 /** Represent a Scio pipeline result. */
-class ScioResult private[scio] (val internal: PipelineResult, val context: ScioContext) {
+abstract class ScioResult private[scio] (val internal: PipelineResult) {
 
-  private val logger = LoggerFactory.getLogger(this.getClass)
+  /** Get a Beam runner specific result. */
+  def as[T <: RunnerResult : ClassTag]: T =
+    ScioUtil.classOf[T].getConstructor(classOf[PipelineResult]).newInstance(internal)
```

Comment:
Should we have some error handling (log message) here - in case of exception from `getCNTR` or `newInstance`?